libbpf-rs/src/ringbuf.rs

```diff
@@ -6,82 +6,128 @@ use std::time::Duration;
-{
-    ptr: *mut libbpf_sys::ring_buffer,
-    ringbufs: Vec<Box<F>>,
+#[derive(Default)]
+pub struct RingBufferBuilder {
+    fd_callbacks: Vec<(i32, Box<Callback>)>,
 }
 
-impl<F> RingBufferManager<F>
-where
-    F: FnMut(&[u8]) -> i32 + 'static,
-{
+impl RingBufferBuilder {
+    pub fn new() -> Self {
+        RingBufferBuilder {
+            fd_callbacks: vec![],
+        }
+    }
+
     /// Add a new ringbuf `map` and associated `callback` to this ring buffer
     /// manager. The callback should take one argument, a slice of raw bytes,
     /// and return an i32.
     ///
-    /// Non-zero return values will stop ring buffer consumption early.
+    /// Non-zero return values in the callback will stop ring buffer consumption early.
+    ///
+    /// Callback is defined as:
+    /// ```rust
+    /// type Callback = fn(&[u8]) -> i32;
+    /// ```
     ///
     /// The callback provides a raw byte slice. You may find libraries such as
     /// [`plain`](https://crates.io/crates/plain) helpful.
-    pub fn add_ringbuf(&mut self, map: &Map, callback: F) -> Result<&mut Self> {
-        let callback = Box::new(callback);
+    pub fn add(&mut self, map: &Map, callback: Callback) -> Result<&mut Self> {
+        if map.map_type() != MapType::RingBuf {
+            return Err(Error::InvalidInput("Must use a RingBuf map".into()));
+        }
+        self.fd_callbacks.push((map.fd(), Box::new(callback)));
+        Ok(self)
+    }
+
+    /// Build a new [`RingBuffer`]. Must have added at least one ringbuf.
+    pub fn build(&self) -> Result<RingBuffer<Callback>> {
+        let mut ringbufs = vec![];
+        let mut ptr: *mut libbpf_sys::ring_buffer = ptr::null_mut();
         let c_sample_cb: libbpf_sys::ring_buffer_sample_fn = Some(Self::call_sample_cb);
-        let sample_cb_ptr = Box::into_raw(callback);
 
-        if self.ptr.is_null() {
-            // Allocate a new ringbuf manager
-            self.ptr = unsafe {
-                libbpf_sys::ring_buffer__new(
-                    map.fd(),
-                    c_sample_cb,
-                    sample_cb_ptr as *mut _,
-                    std::ptr::null_mut(),
-                )
-            };
-
-            // Handle errors
-            let err = unsafe { libbpf_sys::libbpf_get_error(self.ptr as *const _) };
-            if err != 0 {
-                return Err(Error::System(err as i32));
-            }
-        } else {
-            // Add a ringbuf to the existing ringbuf manager
-            let err = unsafe {
-                libbpf_sys::ring_buffer__add(
-                    self.ptr,
-                    map.fd(),
-                    c_sample_cb,
-                    sample_cb_ptr as *mut _,
-                )
-            };
-
-            // Handle errors
-            if err != 0 {
-                return Err(Error::System(err as i32));
+        for (fd, callback) in self.fd_callbacks.clone() {
+            let sample_cb_ptr = Box::into_raw(callback);
+            if ptr.is_null() {
```

Comment:
Instead of adding a special case in the loop, could do:
```
if ringbugs.is_empty():
  return early

ptr = ring_buffer__new()

for cb in cbs:
  ...
```

Reply:
I think this would end up being uglier since ring_buffer__new() actually includes an implicit ring_buffer__add(). Would end up with the same amount of code in the end but with some extra logic to skip over the first callback in the vec.

Reply:
Ah ok. Missed that bit.